### ISSUE TITLE
Incorrect determination of an unused using directive

### ISSUE THREAD

**Issue by GeorgeAlexandria** [dotnet/roslyn#30505]
Hi,

In the **Visual Stuio 15.8.1** (I didn't find that the issue below was fixed at the latest versions), If using directive is used only in the xml doc comments visual studio considers that it's unused and can be fade out. So after formatting a document/selection it will remove this using(fade out is enable) and it will lost a symbol of this usages in the xml doc comment.

Before formatting:
![fadeout](https://user-images.githubusercontent.com/13402478/46917289-13f78000-cfce-11e8-8670-2a8efce8aa99.jpg)
And after:
![fadeout2](https://user-images.githubusercontent.com/13402478/46917324-7486bd00-cfce-11e8-9263-57e4042383f7.jpg)

If this issue should be moved to the [Visual Studio Community](https://developercommunity.visualstudio.com/spaces/8/index.html) instead of Roslyn repository, let me know about it.

**Comment by svick** [comment]
Strange, if I'm reading https://github.com/dotnet/roslyn/issues/2773, https://github.com/dotnet/roslyn/pull/26425 and the git history right, this should not happen anymore in VS 15.8.

**Comment by GeorgeAlexandria** [comment]
@svick, thanks for the links, I read them and was surprised, that it's still not working.

So I tried a couple of cases and found that using directives consider references from xml doc comment which are stayed inside a namespace in a file, and ignore references which are placed outer of namespace as pointed out in the screenshots above.

**Comment by CyrusNajmabadi** [comment]
> So I tried a couple of cases and found that using directives consider references from xml doc comment which are stayed inside a namespace in a file, and ignore references which are placed outer of namespace as pointed out in the screenshots above.

I don't believe there are namespace-level doc comments are allowed in the language.  So this behavior doesn't particularly surprise me.

**Comment by GeorgeAlexandria** [comment]
> I don't believe there are namespace-level doc comments are allowed in the language

@CyrusNajmabadi, What do you mean under "allowed"?! This example can be compiled and visual studio supports navigation and all xml doc features. Maybe this xml doc comment will not be included to the generate file, but it doesn't matter in the scope of this issue.

I tried more cases and find that it's still not work inside the namespace if xml doc comment doesn't attach to type declaration or something nested:

![fadeout3](https://user-images.githubusercontent.com/13402478/46920766-b9751880-cffb-11e8-9e43-0c89cd78fb92.jpg)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, What do you mean under "allowed"

I mean, it's not a supported C# concept.  There is no such thing as a namespace-level doc comment :)

You should be able to observe that by actually compilin the code.  None of the namespace comments will be emitted afaik.

Note: i get warnings here:

![image](https://user-images.githubusercontent.com/4564579/46920829-ccaac880-cfa8-11e8-9f62-a07db9f8fa85.png)

![image](https://user-images.githubusercontent.com/4564579/46920830-d2081300-cfa8-11e8-95a5-48324df38319.png)

**Comment by CyrusNajmabadi** [comment]
> I tried more cases and find that it's still not working inside the namespace if xml doc comment doesn't attach to type declaration or something nested:

Right.  That's as per how the language works.  Xml-doc comments are only allowed on things like types/members.  They don't have any meaning for other constructs (like statements, usings/imports, namespaces, 

> Maybe this xml doc comment will not be included to the generate file, but it doesn't matter in the scope of this issue.

It likely matters because, as far as the language is concerned, this using is *not* used.  :)  Since it never actually touched that comment for any purpose, removing the using had no effect on the end compile.  So the using is safe to remove.

**Comment by CyrusNajmabadi** [comment]
> This example can be compiled and visual studio supports navigation and all xml doc features

Sure.  In the IDE we don't go out of our way to not support this.  But it's still not something the language, using, or compiler actually care about.  From the language's perspective it's the same as if you did:

```c#
//  i am a normal comment.  <see cref="Whatever"/> doesn't meany anything here.
```

Note: it might be hte case that in the *IDE* in the future we'll recognize comments like that and offer highlighting and whatnot.  But as far as the language is concerned, the comment is just a comment, and it has no impact on things like usings.  Cheers!

**Comment by GeorgeAlexandria** [comment]
> //  i am a normal comment.  <see cref="Whatever"/> doesn't meany anything here.
> Note: it might be hte case that in the IDE in the future we'll recognize comments like that and offer highlighting and whatnot

@CyrusNajmabadi, maybe you will be surprised, but it's great work at the current moment, of course if you convert this comment to xml doc!) (and this trick is often used to quickly navigate to a symbol under a comment)

![fadeout4](https://user-images.githubusercontent.com/13402478/46921049-381f8500-cfff-11e8-9e3d-dc9e50eb8e3e.jpg)

By the way, using directive still not consider the usages under a comment inside a method.

> It likely matters because, as far as the language is concerned, this using is not used. :) Since it never actually touched that comment for any purpose, removing the using had no effect on the end compile

It's never touched by compile, yes, it's true, but it actually can be "touched" by IDE to highlight, navigate and many other actions which are really helpful. And it's a sad that this things|actions can be lost because compiler doesn't consider it as usages.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, maybe you will be surprised, but it's great work at the current moment

My example is with a *normal* `//` comment, not a `///` comment :)

**Comment by CyrusNajmabadi** [comment]
> By the way, using directive still not consider the usages under a comment inside a method.

Right.  For the same reason i mentioned above.  The using has no impact on what is compiled by the program, so the language/compiler consider it unused, and so it can be removed.

> It's never touched by compile, yes, it's true, but it actually can be "touched" by IDE to highlight, navigate and many other actions which are really helpful. 

Sure.  But that's true of *all* usings.  I mean, if i have `using static Console`, then someone might use that to navigate to 'Console'.  But we'd still want to remove that 'using' if it was otherwise unused in the file.  :)

I think the goal of "remove unused using" has been to remove anything that's not actually necessary.  In this case, the compiler/IDE even warn that your comment is not something valid and is misplaced.  I don't think there has ever been an expectation that invalidly-placed doc comments would ever get special treatment.
